lib/internal/url.js

```diff
@@ -147,15 +143,120 @@ class URLContext {
+
+  // https://heycam.github.io/webidl/#dfn-default-iterator-object
+  constructor(target, kind) {
+    this.#target = target;
+    this.#kind = kind;
+    this.#index = 0;
+  }
+
+  next() {
+    if (typeof this !== 'object' || this === null || !(#target in this))
+      throw new ERR_INVALID_THIS('URLSearchParamsIterator');
+
+    const index = this.#index;
+    const values = getURLSearchParamsList(this.#target);
+    const len = values.length;
+    if (index >= len) {
+      return {
+        value: undefined,
+        done: true,
+      };
+    }
+
+    const name = values[index];
+    const value = values[index + 1];
+    this.#index = index + 2;
+
+    let result;
+    if (this.#kind === 'key') {
+      result = name;
+    } else if (this.#kind === 'value') {
+      result = value;
+    } else {
+      result = [name, value];
+    }
+
+    return {
+      value: result,
+      done: false,
+    };
+  }
+
+  [inspect.custom](recurseTimes, ctx) {
+    if (!this || typeof this !== 'object' || !(#target in this))
+      throw new ERR_INVALID_THIS('URLSearchParamsIterator');
+
+    if (typeof recurseTimes === 'number' && recurseTimes < 0)
+      return ctx.stylize('[Object]', 'special');
+
+    const innerOpts = { ...ctx };
+    if (recurseTimes !== null) {
+      innerOpts.depth = recurseTimes - 1;
+    }
+    const index = this.#index;
+    const values = getURLSearchParamsList(this.#target);
+    const output = ArrayPrototypeReduce(
+      ArrayPrototypeSlice(values, index),
+      (prev, cur, i) => {
+        const key = i % 2 === 0;
+        if (this.#kind === 'key' && key) {
+          ArrayPrototypePush(prev, cur);
+        } else if (this.#kind === 'value' && !key) {
+          ArrayPrototypePush(prev, cur);
+        } else if (this.#kind === 'key+value' && !key) {
+          ArrayPrototypePush(prev, [values[index + i - 1], cur]);
+        }
+        return prev;
+      },
+      [],
+    );
+    const breakLn = StringPrototypeIncludes(inspect(output, innerOpts), '\n');
+    const outputStrs = ArrayPrototypeMap(output, (p) => inspect(p, innerOpts));
+    let outputStr;
+    if (breakLn) {
+      outputStr = `\n  ${ArrayPrototypeJoin(outputStrs, ',\n  ')}`;
+    } else {
+      outputStr = ` ${ArrayPrototypeJoin(outputStrs, ', ')}`;
+    }
+    return `${this[SymbolToStringTag]} {${outputStr} }`;
+  }
 }
 
+// https://heycam.github.io/webidl/#dfn-iterator-prototype-object
+delete URLSearchParamsIterator.prototype.constructor;
+ObjectSetPrototypeOf(URLSearchParamsIterator.prototype, IteratorPrototype);
+
+ObjectDefineProperties(URLSearchParamsIterator.prototype, {
+  [SymbolToStringTag]: { __proto__: null, configurable: true, value: 'URLSearchParams Iterator' },
+  next: kEnumerableProperty,
+});
+
+
 class URLSearchParams {
-  [searchParams] = [];
+  #searchParams = [];
 
   // "associated url object"
-  [context] = null;
+  #context = null;
+
+  static {
+    setURLSearchParamsContext = (obj, ctx) => {
+      obj.#context = ctx;
```

Comment:
I was looking for a solution like this. Really good! Amazing.

Reply:
Found out about this only recently myself :)